### PR TITLE
[mds-stream] Add Kafka Partition key support to KafkaStreamProducer

### PR DESCRIPTION
**Pull request by avatarneil**
## 📚 Purpose
Implements an option for adding a partition key to KafkaStreamProducers, which will in turn take a hash of the value of that property on each message, and determine what partition should be used within Kafka. This enables horizontal scaling of our Kafka consumers, as producers can write to multiple topics, and ensures that messages are processed in-order depending on the partition key.

## 📦 Impacts:
mds-stream

## ✅ PR Checklist
- [ ] Test in dev cluster

### CODE REVIEW

File: packages/mds-stream/kafka/stream-producer.ts [hunk 1]
```diff
@@ -64,9 +74,15 @@ export const KafkaStreamProducer = <TMessage>(
...
       if (isDefined(producer)) {
         const messages = (isArray(message) ? message : [message]).map(msg => {
-          return { value: JSON.stringify(msg) }
-        })
+          /**
+           * Gets the key to partition on (e.g. a device_id) if partitionKey is set in the options.
+           * By default, this uses a hash of the key to determine what partition it should land on.
+           * If no partitionKey is specified in the options, none will be used.
+           */
+          const getKey = () => (partitionKey ? { key: msg[partitionKey] } : {})
```
Comment (drtyh2o) [inline packages/mds-stream/kafka/stream-producer.ts:82]: I think this function should be declare outside of the `map` and accept `msg` as a parameter so it’s not declared for each message.

File: packages/mds-types/utils.ts [hunk 2]
```diff
@@ -61,3 +60,20 @@ export type Json = Nullable<JsonValue>
 /* eslint-disable @typescript-eslint/no-explicit-any */
 export type AnyFunction<A = any> = (...args: any[]) => A
 export type AnyConstructor<A = object> = new (...args: any[]) => A
+
+/** Keys in T that are also in U */
+export type SetIntersection<T extends keyof any, U extends keyof any> = T extends U ? T : never
```
Comment (drtyh2o) [inline packages/mds-types/utils.ts:69]: Fancy pants.

**Review by drtyh2o — APPROVED** [review]
Minor closure nit.

**Comment by drtyh2o** [comment]
How do we set the number of partitions for a new/existing topic?

**Review by evanxadkins — APPROVED** [review]

File: packages/mds-stream/kafka/stream-producer.ts [hunk 3]
```diff
@@ -64,9 +74,15 @@ export const KafkaStreamProducer = <TMessage>(
...
 
+          return { value: JSON.stringify(msg), ...getKey() }
```
Comment (nessur) [inline packages/mds-stream/kafka/stream-producer.ts:84]:
```suggestion
          const key = partitionKey ? { key: msg[partitionKey] } : {}
          return { value: JSON.stringify(msg), ...key }
```

**Pull request merged**